esp/services/esdl_svc_engine/esdl_binding.cpp

```diff
@@ -551,6 +645,75 @@ static inline bool isPublishedQuery(EsdlMethodImplType implType)
     return (implType==EsdlMethodImplRoxie || implType==EsdlMethodImplWsEcl);
 }
 
+void EsdlServiceImpl::handleFeatureAuthorization(IEspContext& context, IEsdlDefMethod& esdlDef)
+{
+    auto method = esdlDef.queryName();
+    auto bindingMethodAccessMap = m_bindingAccessMaps.getValue(method);
+    MethodAccessMap* accessMap = nullptr;
+#ifdef DEBUG
+    const char* accessMapSource;
+#endif
+
+    if (bindingMethodAccessMap && bindingMethodAccessMap->get())
+    {
+        // The existence of an access map from the binding gives the binding precedence
+        // over the ESDL definition.
+        accessMap = bindingMethodAccessMap->get();
+#ifdef DEBUG
+        accessMapSource = "binding";
+#endif
+    }
+    else
+    {
+        // The absence of an access map from the binding defers the security specification
+        // to the ESDL definition.
+        accessMap = &const_cast<MethodAccessMap&>(esdlDef.queryAccessMap());
+#ifdef DEBUG
+        accessMapSource = "definition";
+#endif
+    }
+
+    if (accessMap->ordinality() > 0)
+    {
+        if (!context.validateFeaturesAccess(*accessMap, false))
+        {
+            StringBuffer features;
+            HashIterator iter(*accessMap);
+            int index = 0;
+            ForEach(iter)
+            {
+                IMapping &cur = iter.query();
+                const char * key = (const char *)cur.getKey();
+                features.appendf("%s%s:%s", (index++ == 0 ? "" : ", "), key, getSecAccessFlagName(*accessMap->getValue(key)));
+            }
+            const char * user = context.queryUserId();
+            throw MakeStringException(-1, "%s::%s access denied for user '%s' - Method requires: '%s'.", context.queryServiceName(nullptr), method, !isEmptyString(user) ? user : "Anonymous", features.str());
+        }
+#ifdef DEBUG
+        else
+        {
+            StringBuffer features;
```

Comment:
Might want to check the log level before calculating all of this if ESPLOG(LogMax will just discard the entry and make the work useless.